### PR TITLE
Registration repository

### PR DESCRIPTION
**Pull request by jaynabonne**

### CODE REVIEW

File: app/uk/gov/hmrc/trusts/controllers/SubmissionDraftController.scala [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+              val sectionPath = JsPath() \ sectionKey
+
+              val thingsToDo = List(
+                prunePath(sectionPath),
+                JsPath.json.update {
+                  sectionPath.json.put(Json.toJson(incomingDraftData.data))
+                },
+                setStatus(incomingDraftData.status)
+              ) ++ setRegistrationSections(incomingDraftData.registrationPieces)
```
Comment (adamconder) [inline app/uk/gov/hmrc/trusts/controllers/SubmissionDraftController.scala:132]: What does this do? :D

File: app/uk/gov/hmrc/trusts/controllers/SubmissionDraftController.scala [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  def setSection(draftId: String, sectionKey: String): Action[JsValue] = identify.async(parse.json) {
+    implicit request => {
+      request.body.validate[RegistrationSubmissionDraftData] match {
+        case JsSuccess(draftData, _) =>
+          submissionRepository.getDraft(draftId, request.identifier).flatMap(
+            result => {
+              val draft: RegistrationSubmissionDraft = result match {
+                case Some(draft) => draft
+                case None => RegistrationSubmissionDraft(draftId, request.identifier, localDateTimeService.now, Json.obj(), None, Some(true))
+              }
```
Comment (adamconder) [inline app/uk/gov/hmrc/trusts/controllers/SubmissionDraftController.scala:46]: There's a lot going on in this controller method, is there a way we can refactor this into single responsible functions?

File: app/uk/gov/hmrc/trusts/controllers/SubmissionDraftController.scala (hunk 1 above)
Comment (jaynabonne) [reply app/uk/gov/hmrc/trusts/controllers/SubmissionDraftController.scala:132]: It is updating three things: the Json for the section (typically user answers, private to individual front end), the status for the section (task list in trusts-frontend), and then all the individual registration section pieces, which come in as an array (submitted data in trusts-frontend). Eventually, there will be the print page rows as well.

File: app/uk/gov/hmrc/trusts/controllers/SubmissionDraftController.scala (hunk 2 above)
Comment (jaynabonne) [reply app/uk/gov/hmrc/trusts/controllers/SubmissionDraftController.scala:46]: It used to be that way, but there was a lot of chatter between the front and back end. If we don't mind having four API calls to the backend per page submit to send all the different pieces, then I can make it be that way. This felt simpler in terms of making sure the data was consistent (e.g. status and reg pieces actually match the user answers).

Comment (jaynabonne) [reply app/uk/gov/hmrc/trusts/controllers/SubmissionDraftController.scala:46]: By doing it all at once, it keeps an update atomic. Splitting the different pieces out risks the data becoming inconsistent.

**Pull request merged**